avalanche/training/supervised/expert_gate.py

```diff
@@ -0,0 +1,293 @@
+        plugins: Optional[List[SupervisedPlugin]] = None,
+        evaluator: EvaluationPlugin = default_evaluator,
+        eval_every=-1,
+        ae_train_mb_size=1,
+        ae_train_epochs=2,
+        ae_lr=1e-3,
+        **base_kwargs
+    ):
+        """Init.
+
+        :param model: PyTorch model.
+        :param optimizer: PyTorch optimizer.
+        :param criterion: loss function.
+        :param train_mb_size: mini-batch size for training.
+        :param train_epochs: number of training epochs.
+        :param eval_mb_size: mini-batch size for eval.
+        :param device: PyTorch device where the model will be allocated.
+        :param plugins: (optional) list of StrategyPlugins.
+        :param evaluator: (optional) instance of EvaluationPlugin for logging
+            and metric computations. None to remove logging.
+        :param eval_every: the frequency of the calls to `eval` inside the
+            training loop. -1 disables the evaluation. 0 means `eval` is called
+            only at the end of the learning experience. Values >0 mean that
+            `eval` is called every `eval_every` epochs and at the end of the
+            learning experience.
+        :param base_kwargs: any additional
+            :class:`~avalanche.training.BaseTemplate` constructor arguments.
+        """
+        # Check that the model has the correct architecture.
+        assert isinstance(
+            model, 
+            ExpertGate), "ExpertGateStrategy requires an ExpertGate model."
+
+        expertgate = _ExpertGatePlugin()
+
+        if plugins is None:
+            plugins = [expertgate]
+        else:
+            plugins += [expertgate]
+
+        self.ae_train_mb_size = ae_train_mb_size
+        self.ae_train_epochs = ae_train_epochs
+        self.ae_lr = ae_lr
+
+        super().__init__(
+                model=model,
+                optimizer=optimizer,
+                criterion=criterion,
+                train_mb_size=train_mb_size,
+                train_epochs=train_epochs,
+                eval_mb_size=eval_mb_size,
+                device=device,
+                plugins=plugins,
+                evaluator=evaluator,
+                eval_every=eval_every,
+                **base_kwargs
+                )
+
+
+class _ExpertGatePlugin(SupervisedPlugin):
+
+    def __init__(self):
+        super().__init__()
+        self.alpha = 0.01
+        self.temp = 2
+        self.lwf_plugin = LwFPlugin(self.alpha, self.temp)
+
+    def before_training_exp(self, 
+                            strategy: "SupervisedTemplate", 
+                            *args, 
+                            **kwargs):
+        super().before_training_exp(strategy, *args, **kwargs)
+
+        task_label = strategy.experience.task_label
+
+        # Always remove the LWF plugin before every experience
+        # If needed, it will be appended later
+        if (self.lwf_plugin in strategy.plugins):
+            strategy.plugins.remove(self.lwf_plugin)
```

Comment:
why do you remove lwf?

Reply:
From my understanding, if a plugin is present then it will be used when training. ExpertGate calls for using LwF only when task relatedness meets a threshold. So, before every experience, I remove the LwF plugin, in case the model used doesn't meet the threshold. 

I add the LwF plugin back, if it meets the threshold, in lines 128-132. If there is a better way to "suppress" plugins that I have missed out on, it would make more sense. I personally don't think what I have done is the cleanest code either.